File: backend/analyze_ppt.py
#!/usr/bin/env python3
"""
Script para analizar la estructura de un archivo PPTX
Extrae dimensiones, layouts, posiciones de elementos, etc.
"""

try:
    from pptx import Presentation
    from pptx.util import Inches, Pt
except ImportError:
    print("❌ python-pptx no está instalado")
    print("   Instalar con: pip install python-pptx")
    exit(1)

import os

def emu_to_inches(emu):
    """Convierte EMU (English Metric Units) a pulgadas"""
    return emu / 914400

def emu_to_cm(emu):
    """Convierte EMU a centímetros"""
    return emu_to_inches(emu) * 2.54

def analyze_ppt(filepath):
    print(f"📊 Analizando: {filepath}\n")
    
    if not os.path.exists(filepath):
        print(f"❌ Archivo no encontrado: {filepath}")
        return
    
    try:
        prs = Presentation(filepath)
        
        # Dimensiones del slide
        print("📐 DIMENSIONES DEL SLIDE:")
        width_inches = emu_to_inches(prs.slide_width)
        height_inches = emu_to_inches(prs.slide_height)
        width_cm = emu_to_cm(prs.slide_width)
        height_cm = emu_to_cm(prs.slide_height)
        
        print(f"   Ancho:  {width_inches:.2f}\" ({width_cm:.2f} cm)")
        print(f"   Alto:   {height_inches:.2f}\" ({height_cm:.2f} cm)")
        print(f"   Proporción: {width_inches/height_inches:.2f}:1\n")
        
        # Número de slides
        print(f"📄 TOTAL SLIDES: {len(prs.slides)}\n")
        
        # Layouts disponibles
        print(f"🎨 LAYOUTS DISPONIBLES: {len(prs.slide_layouts)}")
        for idx, layout in enumerate(prs.slide_layouts):
            print(f"   [{idx}] {layout.name}")
        print()
        
        # Analizar cada slide
        for slide_idx, slide in enumerate(prs.slides):
            print(f"\n{'='*60}")
            print(f"📄 SLIDE {slide_idx + 1}")
            print(f"   Layout: {slide.slide_layout.name}")
            print(f"{'='*60}")
            
            # Analizar shapes en el slide
            print(f"\n   🔲 ELEMENTOS EN EL SLIDE: {len(slide.shapes)}\n")
            
            for shape_idx, shape in enumerate(slide.shapes):
                print(f"   [{shape_idx}] {shape.shape_type} - {shape.name}")
                
                # Posición y dimensiones
                x_inches = emu_to_inches(shape.left)
                y_inches = emu_to_inches(shape.top)
                w_inches = emu_to_inches(shape.width)
                h_inches = emu_to_inches(shape.height)
                
                print(f"       Posición: X={x_inches:.2f}\", Y={y_inches:.2f}\"")
                print(f"       Tamaño:   W={w_inches:.2f}\", H={h_inches:.2f}\"")
                
                # Si es texto
                if hasattr(shape, 'text'):
                    text = shape.text.strip()
                    if text:
                        preview = text[:50] + "..." if len(text) > 50 else text
                        print(f"       Texto: \"{preview}\"")
                        
                        if hasattr(shape, 'text_frame'):
                            for para in shape.text_frame.paragraphs:
                                if para.runs:
                                    run = para.runs[0]
                                    if run.font.size:
                                        size_pt = run.font.size.pt
                                        print(f"       Fuente: {size_pt:.0f}pt", end="")
                                        if run.font.bold:
                                            print(" BOLD", end="")
                                        if run.font.color.rgb:
                                            print(f", Color: #{run.font.color.rgb}", end="")
                                        print()
                
                # Si es imagen
                if shape.shape_type == 13:  # PICTURE
                    print(f"       🖼️ IMAGEN detectada")
                    if hasattr(shape, 'image'):
                        print(f"       Formato: {shape.image.content_type}")
                
                print()
        
        print(f"\n{'='*60}")
        print("✅ Análisis completado")
        print(f"{'='*60}\n")
        
        # Resumen de configuración recomendada
        print("💡 CONFIGURACIÓN RECOMENDADA PARA PPTXGENJS:")
        print(f"\nconst pptx = new PptxGenJS();")
        print(f"// NO establecer layout - usar dimensiones por defecto 10x7.5")
        print(f"\nconst slide = pptx.addSlide();")
        print(f"\n// Para mantener proporciones del archivo original:")
        print(f"// Width: {width_inches:.2f}\" Height: {height_inches:.2f}\"")
        print(f"// (PptxGenJS default es 10x7.5, muy similar)\n")
        
    except Exception as e:
        print(f"❌ Error al analizar: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    filepath = r"c:\Users\migduran\Documents\nuevo ooh\backend\REPORTE FACTURACIÓN BASE.pptx"
    analyze_ppt(filepath)

File: backend/analyze_ppt.js
// Script para analizar un PPT existente y obtener dimensiones
const PptxGenJS = require('pptxgenjs');
const fs = require('fs');
const path = require('path');

async function analyzePPT() {
  try {
    console.log('📊 Analizando REPORTE FACTURACIÓN BASE.pptx...\n');
    
    // Leer el archivo
    const filePath = path.join(__dirname, 'REPORTE FACTURACIÓN BASE.pptx');
    
    if (!fs.existsSync(filePath)) {
      console.error('❌ Archivo no encontrado:', filePath);
      return;
    }
    
    console.log('✅ Archivo encontrado');
    console.log('📏 Tamaño:', fs.statSync(filePath).size, 'bytes\n');
    
    // Crear un PPT de prueba con las dimensiones estándar
    const pptx = new PptxGenJS();
    
    // Mostrar dimensiones por defecto
    console.log('📐 DIMENSIONES ESTÁNDAR DE PPTXGENJS:');
    console.log('   Ancho: 10 pulgadas (25.4 cm)');
    console.log('   Alto: 7.5 pulgadas (19.05 cm)');
    console.log('   Proporción: 4:3\n');
    
    // Dimensiones recomendadas para elementos
    console.log('📏 DIMENSIONES RECOMENDADAS PARA ELEMENTOS:');
    console.log('\n🖼️ IMÁGENES (para slide 10"x7.5"):');
    console.log('   Imagen grande (izquierda):');
    console.log('     - Ancho: 4.5" (11.43 cm)');
    console.log('     - Alto: 4.5" (11.43 cm)');
    console.log('     - X: 0.5" (desde borde izquierdo)');
    console.log('     - Y: 1.5" (desde arriba)');
    console.log('\n   Imágenes pequeñas (derecha, stack):');
    console.log('     - Ancho: 4" (10.16 cm)');
    console.log('     - Alto: 2" (5.08 cm)');
    console.log('     - X: 5.5" (desde borde izquierdo)');
    console.log('     - Y1: 1.5" (imagen superior)');
    console.log('     - Y2: 3.75" (imagen inferior)');
    console.log('\n📝 TEXTO:');
    console.log('   Título (Dirección):');
    console.log('     - X: 0.5", Y: 0.5"');
    console.log('     - Tamaño: 24pt, Bold');
    console.log('     - Color: #CC0000 (Rojo ABI)');
    console.log('\n   Subtítulo (Ciudad):');
    console.log('     - X: 0.5", Y: 1"');
    console.log('     - Tamaño: 14pt');
    console.log('     - Color: #003366 (Azul)');
    console.log('\n   Vigencia:');
    console.log('     - X: 5.5", Y: 6"');
    console.log('     - Tamaño: 12pt');
    console.log('\n   REF (Proveedor):');
    console.log('     - X: 0.5", Y: 6.5"');
    console.log('     - Tamaño: 12pt');
    
    console.log('\n🎨 COLORES ABI:');
    console.log('   Rojo: #CC0000');
    console.log('   Oro: #D4A574');
    console.log('   Azul: #003366');
    
    console.log('\n💡 NOTAS IMPORTANTES:');
    console.log('   - PptxGenJS usa pulgadas (inches) como unidad');
    console.log('   - 1 pulgada = 2.54 cm');
    console.log('   - Coordenadas (X,Y) empiezan en esquina superior izquierda (0,0)');
    console.log('   - Ancho del slide: 10"');
    console.log('   - Alto del slide: 7.5"');
    console.log('   - Margen recomendado: 0.5" en todos los lados\n');
    
    // Información sobre cómo copiar el diseño del archivo base
    console.log('🔧 PARA USAR DISEÑO DEL ARCHIVO BASE:');
    console.log('   PptxGenJS no puede copiar diseños de archivos existentes.');
    console.log('   Opciones:');
    console.log('   1. Recrear el diseño manualmente con las dimensiones correctas');
    console.log('   2. Usar una plantilla .potx como base (si está disponible)');
    console.log('   3. Definir un master slide personalizado en PptxGenJS\n');
    
  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

analyzePPT();
